### PR TITLE
Limit user access to VDI by AWS credit

### PR DESCRIPTION
**Pull request by corentindrouet**
Add An aws credit limit, 

Add getUserHistory for searching the time use of all machines of a user.
Add getUserCredit using getUserHistory for calculate the global credit used by a user.
Backend return a 402 error if the user exceeded his credit limit when trying to opening a vm.
It's possible to set this limit in Configuration.

### CODE REVIEW

File: api/controllers/AppController.js [hunk 1]
```diff
@@ -142,7 +145,11 @@ module.exports = {
           });
       })
       .catch((err) => {
-        return res.negotiate(err);
+        if (err !== 'Exceeded credit') {
```
Comment (dynamiccast) [inline api/controllers/AppController.js]: This condition would be better the other way around:

```
if (err === 'Exceeded credit') {
  return res.send(402, err);
} else {
  return res.negotiate(err);
}
```

File: api/controllers/MachineController.js [hunk 2]
```diff
@@ -51,10 +51,18 @@ module.exports = {
    * @method users
    */
   users(req, res) {
-    Machine.find()
+    MachineService.getMachineForUser(req.user)
       .then((machines) => {
-        return res.ok(machines);
+        let machineArray = [];
+        machineArray.push(machines);
+        return res.ok(machineArray);
```
Comment (dynamiccast) [inline api/controllers/MachineController.js]: You can reply the array directly:

```
return res.ok([
  machines
]);
```

File: api/drivers/aws/driver.js [hunk 3]
```diff
@@ -89,6 +90,44 @@ class AWSDriver extends Driver {
...
+                      'ap-southeast-1': 'Asia Pacific (Singapore)',
+                      'ap-southeast-2': 'Asia Pacific (Sydney)',
+                      'ap-northeast-1': 'Asia Pacific (Tokyo)',
+                      'eu-central-1': 'EU (Frankfurt)',
+                      'eu-west-1': 'EU (Ireland)'
+                    };
+                    let keyPrice = [];
+                    request('https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/index.json', (err, res, body) => {
+                      if (err) {
+                        res.negotiate(err);
```
Comment (dynamiccast) [inline api/drivers/aws/driver.js]: All terminal method should be returned

File: api/drivers/aws/driver.js [hunk 4]
```diff
@@ -89,6 +90,44 @@ class AWSDriver extends Driver {
...
+                      } else {
+                        Promise.resolve(JSON.parse(body))
+                          .then((prix) => {
+                            priceObjet = prix;
```
Comment (dynamiccast) [inline api/drivers/aws/driver.js]: Please use english variable name

File: api/drivers/aws/driver.js [hunk 5]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
       });
     });
   }
+
+  /**
+   * Calculate credit used by a user
+   *
+   * @method getUserCredit
+   * @param {user} user The user how have close the session
```
Comment (dynamiccast) [inline api/drivers/aws/driver.js]: @param {Object}
"The user how have close the session" -> "User to calculate credit usage from"

File: api/drivers/aws/driver.js [hunk 6]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+   * @return {Promise[price]}
```
Comment (dynamiccast) [inline api/drivers/aws/driver.js]: price is not a type. Is the promise resolving to a number ?

File: api/drivers/driver.js [hunk 7]
```diff
@@ -74,6 +74,17 @@ class Driver {
   getServer(/* id */) {
     return Promise.reject(new Error('Driver\'s method "getServer" not implemented'));
   }
+
+  /**
+   * Return price
```
Comment (dynamiccast) [inline api/drivers/driver.js]: 'Return price" -> "Calculate credit used by a user"

File: api/drivers/driver.js [hunk 8]
```diff
@@ -74,6 +74,17 @@ class Driver {
...
+   *
+   * @method getUserCredit
+   * @param {User} User model
```
Comment (dynamiccast) [inline api/drivers/driver.js]: User -> Object

File: api/drivers/driver.js [hunk 9]
```diff
@@ -74,6 +74,17 @@ class Driver {
...
+   * @return {Promise[price]}
```
Comment (dynamiccast) [inline api/drivers/driver.js]: price -> number

File: api/drivers/dummy/driver.js [hunk 10]
```diff
@@ -113,7 +113,6 @@ class DummyDriver extends BaseDriver {
       .then(resolve, reject);
     });
   }
-
   destroyMachine(machine) {
```
Comment (dynamiccast) [inline api/drivers/dummy/driver.js:150]: Why did you remove this lime ?

File: api/services/UserService.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ */
+
+/* global History */
+
+const Promise = require('bluebird');
+const moment = require('moment');
+
+/**
+ * Return the credit used with the machine for a user
```
Comment (dynamiccast) [inline api/services/UserService.js]: No idea what this means

File: api/services/UserService.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @param {user, driver} user The user we need for get the history
```
Comment (dynamiccast) [inline api/services/UserService.js]: Please use two "@param"

File: api/services/UserService.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+ * driver The driver used by the user
+ * @return {Promise[[object]]}
```
Comment (dynamiccast) [inline api/services/UserService.js]: To what does this promise resolves ?

File: api/services/UserService.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    return History.find({
+      userId: user.id,
+      machineDriver: driver
+    })
+      .then((allHistories) => {
+        return allHistories.forEach((element) => {
+          var obj = machines.filter((obj) => {
+            return obj.id === element.machineId;
+          });
+          if (!obj[0]) {
```
Comment (dynamiccast) [inline api/services/UserService.js]: Prefer `obj.length` to check if an array if empty.
What is `obj` anyway ?

File: api/controllers/HistoryController.js [hunk 15]
```diff
@@ -25,18 +25,52 @@
...
 
-/* global MachineService, JsonApiService */
+/* global MachineService, JsonApiService, History */
 
 module.exports = {
   create(req, res) {
     req.body = JsonApiService.deserialize(req.body);
 
-    if (req.body.data.attributes['end-date'] === '') {
+    if (req.body.data.attributes.endDate === '') {
```
Comment (Costent) [inline api/controllers/HistoryController.js]: JSONAPI expects snake-case attribute names

File: api/controllers/HistoryController.js [hunk 16]
```diff
@@ -25,18 +25,52 @@
...
+    var data = req.body.data.attributes || {};
+    var id = req.options.id || (req.options.where && req.options.where.id) || req.allParams().id;
+
+    if (req.allParams().id) {
+      data.id = req.allParams().id;
+    }
+    History.update(id, data)
+      .then((records) => {
+        if (!records || !records.length || records.length > 1) {
+          req._sails.log.warn('Unexpected output from `' + History.globalId + '.update`.');
```
Comment (willrnch) [inline api/controllers/HistoryController.js]: Abort the request and return an error in that case

File: api/controllers/HistoryController.js [hunk 17]
```diff
@@ -25,18 +25,52 @@
...
+        if (updatedRecord === undefined) {
+          return res.notFound();
+        }
+
+        return res.ok(updatedRecord);
+      })
+      .then(() => {
+        if (data.endDate !== '') {
+          MachineService.sessionEnded(req.user);
+        }
```
Comment (willrnch) [inline api/controllers/HistoryController.js:76]: Unhandled case

File: api/drivers/aws/driver.js [hunk 18]
```diff
@@ -195,7 +234,7 @@ class AWSDriver extends Driver {
             const password = res.password;
 
 
-            return ConfigService.get('awsMachineUsername', 'plazaPort')
+            return ConfigService.get('awsMachineUsername', 'plazaPort', 'awsFlavor')
```
Comment (Costent) [inline api/drivers/aws/driver.js]: Is `plazaPort` needed here?

File: api/drivers/aws/driver.js [hunk 19]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+    return new Promise((resolve, reject) => {
+      Promise.props({
+        conf: ConfigService.get('price'),
+        timeHistory: UserService.getUserHistory(user, 'aws')
+      })
+        .then((result) => {
+          jsonPrice = result.conf.price;
+          duration = result.timeHistory;
+          var prod = Object.keys(jsonPrice.products);
+          return duration.forEach((element) => {
```
Comment (willrnch) [inline api/drivers/aws/driver.js]: confusing `return`

File: api/drivers/aws/driver.js [hunk 20]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+            return prod.forEach((elem) => {
```
Comment (willrnch) [inline api/drivers/aws/driver.js]: confusing `return`

File: api/drivers/aws/driver.js [hunk 21]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+                product.push(jsonPrice.products[elem]);
+                return jsonPrice.products[elem];
+              } else {
+                return null;
+              }
+            });
+          });
+        })
+        .then(() => {
+          return product.forEach((element) => {
```
Comment (willrnch) [inline api/drivers/aws/driver.js]: confusing `return`

File: api/drivers/aws/driver.js [hunk 22]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+
+            tmp = jsonPrice.terms.OnDemand[element.sku];
+            tmpKey = Object.keys(tmp);
+            tmp = tmp[tmpKey].priceDimensions;
+            tmpKey = Object.keys(tmp);
+            priceAll.push(tmp[tmpKey].pricePerUnit.USD);
+          });
+        })
+        .then(() => {
+          return duration.forEach((element, index) => {
```
Comment (willrnch) [inline api/drivers/aws/driver.js]: confusing `return`

File: api/drivers/aws/driver.js [hunk 23]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+                return null;
+              }
+            });
+          });
+        })
+        .then(() => {
+          return product.forEach((element) => {
+            var tmp = {};
+            var tmpKey = null;
+
```
Comment (willrnch) [inline api/drivers/aws/driver.js]: What does this do? Comment and/or use lodash

File: api/drivers/aws/driver.js [hunk 24]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+    return new Promise((resolve, reject) => {
+      Promise.props({
+        conf: ConfigService.get('price'),
+        timeHistory: UserService.getUserHistory(user, 'aws')
+      })
+        .then((result) => {
+          jsonPrice = result.conf.price;
+          duration = result.timeHistory;
+          var prod = Object.keys(jsonPrice.products);
```
Comment (willrnch) [inline api/drivers/aws/driver.js]: What does this do? Comment and/or use lodash

File: api/drivers/aws/driver.js [hunk 25]
```diff
@@ -255,6 +295,67 @@ class AWSDriver extends Driver {
...
+            tmpKey = Object.keys(tmp);
+            priceAll.push(tmp[tmpKey].pricePerUnit.USD);
+          });
+        })
+        .then(() => {
+          return duration.forEach((element, index) => {
+            finalPrice += (element.time * priceAll[index]);
+          });
+        })
+        .then(() => {
```
Comment (willrnch) [inline api/drivers/aws/driver.js]: You should use `.return(finalPrice)` or better, retrieve the value from the previous Promise.

File: api/services/MachineService.js [hunk 26]
```diff
@@ -322,7 +330,23 @@ function sessionOpen(user) {
...
+    })
+      .then((creditUsed) => {
+        return User.update({
+          id: user.id
+        }, {
+          credit: creditUsed
+        });
+      })
+      .catch((err) => {
+        return new Error(err);
```
Comment (willrnch) [inline api/services/MachineService.js]: useless

File: api/services/MachineService.js [hunk 27]
```diff
@@ -322,7 +330,23 @@ function sessionOpen(user) {
  * @return {Promise}
  */
 function sessionEnded(user) {
-  return increaseUsersMachineEndDate(user);
+  var ret = increaseUsersMachineEndDate(user);
+  if (_driver.name() === 'aws') {
```
Comment (willrnch) [inline api/services/MachineService.js]: Don't rely on the driver name here. You may want to add a method `isUserCreditSupported` or something.

File: api/services/UserService.js [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+  let duration = [];
+  let machines = [];
+
+  return new Promise((resolve, reject) => {
+    return History.find({
+      userId: user.id,
+      machineDriver: driver
+    })
+      .then((allHistories) => {
+        return allHistories.forEach((element) => {
```
Comment (willrnch) [inline api/services/UserService.js]: Confusing return

File: api/services/UserService.js [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+              listedMachines[0].start = element.startDate;
+            }
+            if (listedMachines[0].end < element.endDate) {
+              listedMachines[0].end = element.endDate;
+            }
+          }
+        });
+      })
+      .then(() => {
+        return machines.forEach((element) => {
```
Comment (willrnch) [inline api/services/UserService.js]: confusing return

File: api/services/UserService.js [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+            type: element.type,
+            time: hoursDiff
+          });
+        });
+      })
+      .then(() => {
+        return resolve(duration);
+      })
+      .catch((err) => {
+        return reject(err);
```
Comment (willrnch) [inline api/services/UserService.js]: useless

File: api/services/UserService.js [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+          var hoursDiff = element.end.diff(element.start, 'hours') + 1;
+
+          duration.push({
+            type: element.type,
+            time: hoursDiff
+          });
+        });
+      })
+      .then(() => {
+        return resolve(duration);
```
Comment (willrnch) [inline api/services/UserService.js]: Where does `duration` comes from ?

File: assets/app/protected/apps/index/controller.js [hunk 32]
```diff
@@ -39,23 +39,28 @@ let App = Ember.Object.extend({
...
           .then(() => {
-          })
-          .catch(() => {
-            this.toast.error('Cannot start application');
-          })
-          .finally(() => {
-            this.set('remoteSession.plazaHasFinishedLoading', true);
+            app.set('state', 'running');
+            app.save()
+              .then(() => {
```
Comment (willrnch) [inline assets/app/protected/apps/index/controller.js]: then what?

File: assets/app/protected/apps/index/controller.js [hunk 33]
```diff
@@ -39,23 +39,28 @@ let App = Ember.Object.extend({
...
+              })
+              .finally(() => {
+                this.set('remoteSession.plazaHasFinishedLoading', true);
+              });
           });
+      })
+      .catch((err) => {
+        if (err === 'Exceeded credit') {
+          this.toast.error('Exceeded credit');
+        }
```
Comment (willrnch) [inline assets/app/protected/apps/index/controller.js:62]: Unhandled case

File: assets/app/protected/apps/index/controller.js [hunk 34]
```diff
@@ -146,9 +151,14 @@ export default Ember.Controller.extend({
             rej();
           }
         })
-      .finally(() => {
-        this.set('isCheckingMachine', false);
-      });
+        .catch((err) => {
+          if (err.errors && err.errors[0].status === '402') {
+            this.toast.error('Credit exceeded');
+          }
```
Comment (willrnch) [inline assets/app/protected/apps/index/controller.js:159]: Unhandled case

File: assets/app/protected/configs/user-right/template.hbs [hunk 35]
```diff
@@ -43,4 +43,7 @@
   {{input type="checkbox" checked=configController.autoLogoff}}
   <label for="cbox2">Logoff after closing VDI</label>
 </div>
-
+<div>
+  {{input type="number" value=configController.creditLimit}}
+  <label for="cbox2">Set a credit limit to users using AWS</label>
```
Comment (willrnch) [inline assets/app/protected/configs/user-right/template.hbs]: Can't we find a way to hide this message is the driver is not AWS?

File: config/env/development.js [hunk 36]
```diff
@@ -64,9 +64,11 @@ module.exports = {
     autoLogoff: false,
     defaultGroup: '',
     awsCache: 0,
+    creditLimit: '',
```
Comment (willrnch) [inline config/env/development.js:67]: Add documentation about this (README)

File: config/env/development.js [hunk 37]
```diff
@@ -64,9 +64,11 @@ module.exports = {
...
 
     iaas: 'manual',
     machinePoolSize: 1,
+    price: {},
```
Comment (willrnch) [inline config/env/development.js]: Don't store this in config

File: tests/unit/drivers/driver.test.js [hunk 38]
```diff
@@ -0,0 +1,133 @@
...
+
+process.env.iaas = 'dummy';
+process.env.machinePoolSize = 3;
+process.env.sessionDuration = 0;
+
+describe('aws driver', () => {
+  describe('getUserCredit', () => {
+    before('clean history database', function(done) {
+      History.query('TRUNCATE TABLE public.history', () => {
+        fs.readFile('./tests/unit/drivers/fakePrice.json', (err, data) => {
```
Comment (willrnch) [inline tests/unit/drivers/driver.test.js]: This is internal to the AWS driver. Should not be handled here.

File: config/env/development.js [hunk 39]
```diff
@@ -64,9 +64,11 @@ module.exports = {
     autoLogoff: false,
     defaultGroup: '',
     awsCache: 0,
+    creditLimit: '',
 
     iaas: 'manual',
     machinePoolSize: 1,
+//    price: {},
```
Comment (dynamiccast) [inline config/env/development.js]: Remaining comment

File: api/drivers/aws/driver.js [hunk 40]
```diff
@@ -107,6 +109,74 @@ class AWSDriver extends Driver {
   }
 
   /**
+   * Get the aws price
+   * It reduce the AwsPrice Object by deleting useless key
```
Comment (dynamiccast) [inline api/drivers/aws/driver.js:116]: Can you precise this method's job is to set `this.awsPrice` to a promise that will resolve to the prices retrieved from AWS ?

File: api/services/UserService.js [hunk 41]
```diff
@@ -0,0 +1,146 @@
...
+            machines.push({
+              id: element.machineId,
+              type: element.machineType,
+              start: [element.startDate],
+              end: [element.endDate]
+            });
+          } else {
+
+            /**
+             * If no history overlap with the current history, it push it to the
```
Comment (dynamiccast) [inline api/services/UserService.js]: overlap -> overlaps
push -> pushes

File: README.md [hunk 42]
```diff
@@ -34,6 +34,7 @@ AWS driver specific
 - AWS_IMAGE (defaults to ami-09e61366, Nanocloud default image)
 - AWS_FLAVOR (defaults to t2.medium) size of the virtual machine
 - AWS_MACHINE_USERNAME (defaults to Administrator) administrator account on the machine
+- CREDIT_LIMIT (defaults unset) set a credit limit to users using aws
```
Comment (dynamiccast) [inline README.md]: Would not it be -> default empty string ?

**Comment by dynamiccast** [comment]
This PR has conflicts that must be resolved

**Comment by dynamiccast** [comment]
If I connect to the VDI with the admin for 2 minutes, then hit back to my profile I can see I used 0 credit. Is that normal ? Should not I have consume just a little bit of credit ?

**Comment by dynamiccast** [comment]
If I set the maximum credit usage to 0 I can still use the VDI.
